README.md
# eus_vive

[![main](https://github.com/knorth55/eus_vive/actions/workflows/main.yml/badge.svg)](https://github.com/knorth55/eus_vive/actions/workflows/main.yml)
[![linter](https://github.com/knorth55/eus_vive/actions/workflows/linter.yaml/badge.svg)](https://github.com/knorth55/eus_vive/actions/workflows/linter.yaml)

Multi robot teleoperation system with Vive/SpaceNav/Oculus/Tablis Cockpit

## Note

This package depends on these branches below:

- For PR2 and  Baxter
  - [knorth55/rviz_camera_stream@master](https://github.com/knorth55/rviz_camera_stream)
  - [knorth55/rviz_textured_sphere@add-padding](https://github.com/knorth55/rviz_textured_sphere/tree/add-padding)
  - [knorth55/vive_ros@use-camera](https://github.com/knorth55/vive_ros/tree/use-camera)

- For Baxter only
  - [knorth55/baxter_interface@baxter-vive-control](https://github.com/knorth55/baxter_interface/tree/baxter-vive-control)
  - [pazeshun/dynamixel_motor@gripper-v6-devel](https://github.com/pazeshun/dynamixel_motor/tree/gripper-v6-devel)

- For Dragon only
  - [knorth55/jsk_aeriao_robot@drone-euslisp](https://github.com/knorth55/jsk_aerial_robot/tree/drone-euslisp)

## Tested Environment

### Build environment

#### Ubuntu 20.04 + ROS Noetic

Steam does not work on Melodic or Kinetic now.

- NVidia driver: `525.105`
- OpenVR: `1.3.22`
- Steam VR: `1.6.10`

### User interface devices

#### Vive

- [x] Arm motion tracking
- [x] Controller button interface
- [x] HMD visual interface
- [x] Vibration interface
- [x] Sound interface

#### Tablis

- [x] Arm motion tracking
- [x] Controller button interface
- [x] HMD visual interface
- [x] Vibration interface
- [x] Sound interface

#### SpaceNav (3D mouse)

- [x] Arm motion tracking
- [x] Controller button interface

#### Oculus

- [x] Arm motion tracking
- [ ] Controller button interface
- [ ] HMD visual interface
- [ ] Vibration interface
- [ ] Sound interface

## Installation

### Dependency installation

#### Install dependencies (for Vive)

```bash
sudo apt-get install --reinstall xserver-xorg-video-intel-hwe-18.04 libgl1-mesa-glx libgl1-mesa-dri xserver-xorg-core
sudo dpkg-reconfigure xserver-xorg
```

#### Install nvidia-driver (for Vive)

```bash
# for melodic, run command below
sudo apt install nvidia-driver-390
# for kinetic, install cuda-9.2 deb (local) manually and run command below.
sudo apt install nvidia-396
```

#### Install `OpenVR`, `steam` and `steamVR` (for Vive)

Follow [here](https://github.com/knorth55/vive_ros)

#### Remove libcurl packages in Steam (for Vive)

```bash
rm -r ~/.steam/steam/ubuntu12_32/steam-runtime/usr/lib/x86_64-linux-gnu/libcurl* 
```

#### Install kodak 4k pro camera and ELP usb camera udev (for Baxter)

```bash
sudo cp udev/99-kodak.rules /etc/udev/rules.d/
sudo udevadm control --reload-rules && sudo udevadm trigger
```

### ROS Workspace build

#### Build `jsk_apc` workspace (Only for baxter users)

```bash
source /opt/ros/$ROS_DISTRO/setup.bash
mkdir ~/jsk_apc_ws/src -p
cd ~/jsk_apc_ws/src
wstool init . https://raw.githubusercontent.com/start-jsk/jsk_apc/master/fc.rosinstall.${ROS_DISTRO}
wstool up
rosdep install -y -r --from-paths .

cd ~/jsk_apc_ws 
catkin config
catkin build
```

#### Build `eus_vive` workspace

```bash
source /opt/ros/$ROS_DISTRO/setup.bash
mkdir ~/vive_ws/src -p
cd ~/vive_ws/src
wstool init . https://raw.githubusercontent.com/knorth55/eus_vive/master/fc.rosinstall

# Only for baxter users
wstool merge https://raw.githubusercontent.com/knorth55/eus_vive/master/baxter.rosinstall
wstool merge https://raw.githubusercontent.com/knorth55/eus_vive/master/baxter.rosinstall.$ROS_DISTRO

# Only for dragon users
wstool merge https://raw.githubusercontent.com/knorth55/eus_vive/master/dragon.rosinstall

wstool up
rosdep install --ignore-src --from-path . -y -r -i

# Only for baxter users
source ~/jsk_apc_ws/devel/setup.bash

cd ~/vive_ws
catkin config
catkin build
```

## How to start

### Start procedure

- Connect Vive HMD and Lighthouse and power on the controller.
- Place Lighthouse B in front of you.
- Start one of launch files below.
- If you don't use HMD, press `Calibrate` button and do arm calibration.
- Press `Enable` button in GUI to start teleoperation.

### Arm calibration (Only for no HMD mode)

The arm calibration is required for no HMD mode in order to measure your position and your arm length.

- Press `Calibrate l/rarm` button in GUI
- Listen to the voice instruction.
- Stretch your left/right arm and Press Trigger after the first instruction.
- Listen to the voice instruction again.
- Fold your left/right arm and Press Trigger after the second instruction.
- If calibration failed, please try again.

### PR2 + Vive

https://user-images.githubusercontent.com/9300063/213145161-4b2acb89-9545-460a-b5cc-d05a111a187b.mp4

#### PR2 + Vive in JSK 73B2 or 610

```bash
rossetip
rossetmaster pr1040
# HMD mode
roslaunch eus_vive pr2_vive.launch
# No HMD mode
roslaunch eus_vive pr2_vive.launch head:=false
```

#### PR2 + Vive in Gazebo

```bash
roslaunch pr2_gazebo pr2_empty_world.launch
roslaunch eus_vive pr2_vive_gazebo.launch
```

### PR2 + Tablis

https://user-images.githubusercontent.com/9300063/208720417-176d698e-3789-42a8-931f-2a9fc9e062a0.mp4

#### PR2 + Tablis in JSK 73B2 or 610

```bash
rossetip
rossetmaster pr1040
roslaunch eus_vive pr2_tablis.launch
```

#### PR2 + Tablis in Gazebo

##### Launch Tablis in Choreonoid

```bash
roscd eus_vive/scripts/tablis
./start-tablis-sim.sh
```

```bash
roscd eus_vive/scripts/tablis
ipython -i tablis_setup.py
hcf.servoOn()
hcf.hc_svc.startHapticsController()
```

##### Launch bridge

```bash
roscd eus_vive/scripts/tablis
./start-bridge-sim.sh
```

##### Launch PR2 in Gazebo

```bash
roslaunch pr2_gazebo pr2_empty_world.launch
roslaunch eus_vive pr2_tablis_gazebo.launch
```

### PR2 + SpaceNav

https://user-images.githubusercontent.com/9300063/213137459-ce5f7075-acee-4a47-934a-97a4296fdc49.mp4

#### PR2 + SpaceNav in JSK 73B2

```bash
rossetip
rossetmaster pr1040
roslaunch eus_vive pr2_spacenav.launch
```

#### PR2 + SpaceNav in Gazebo

```bash
roslaunch pr2_gazebo pr2_empty_world.launch
roslaunch eus_vive pr2_spacenav_gazebo.launch
```

### Baxter + Vive

https://user-images.githubusercontent.com/9300063/213144997-0d224d5c-b462-43a7-b759-0233fff41d94.mp4

#### Baxter + Vive in JSK 73B2

##### Baxter control PC

```bash
rossetip
rossetmaster baxter
roslaunch eus_vive baxter_remote.launch
```

##### Vive PC

```bash
rossetip
rossetmaster baxter
# No HMD mode
roslaunch eus_vive baxter_vive_remote.launch
# for display
roslaunch eus_vive baxter_display_remote.launch
```

#### Baxter + Vive in Gazebo

```bash
roslaunch baxter_gazebo baxter_world.launch
roslaunch eus_vive baxter_vive_gazebo.launch
```

### Baxter + Tablis

https://user-images.githubusercontent.com/9300063/208720482-b7ef0abb-e948-448c-be51-521c0c7af8a2.mp4

#### Baxter + Tablis in JSK 73B2 or 610

```bash
rossetip
rossetmaster baxter
roslaunch eus_vive baxter_tablis.launch
```

#### Baxter + Tablis in Gazebo

##### Launch Tablis in Choreonoid

```bash
roscd eus_vive/scripts/tablis
./start-tablis-sim.sh
```

```bash
roscd eus_vive/scripts/tablis
ipython -i tablis_setup.py
hcf.servoOn()
hcf.hc_svc.startHapticsController()
```

##### Launch bridge

```bash
roscd eus_vive/scripts/tablis
./start-bridge-sim.sh
```

##### Launch Baxter in Choreonoid

```bash
roslaunch baxter_gazebo baxter_world.launch
roslaunch eus_vive baxter_tablis_gazebo.launch
```

### Baxter + SpaceNav

#### Baxter + SpaceNav in JSK 73B2

```bash
rossetip
rossetmaster baxter
roslaunch eus_vive baxter_spacenav.launch
```

#### Baxter + SpaceNav in Gazebo

```bash
roslaunch baxter_gazebo baxter_world.launch
roslaunch eus_vive baxter_spacenav_gazebo.launch
```

### Baxter + MoveIt!

#### Real Robot in JSK 73B2

```bash
rossetip
rossetmaster baxter
roslaunch eus_vive baxter_moveit.launch
```

### Dragon + SpaceNav

https://user-images.githubusercontent.com/9300063/210754022-7df92fd5-4875-4465-9738-7efdf5649093.mp4

#### Dragon + SpaceNav in Gazebo

```bash
roslaunch dragon bringup.launch simulation:=true real_machine:=false headless:=false
roslaunch eus_vive dragon_spacenav_gazebo.launch
```

### JAXON + SpaceNav

https://user-images.githubusercontent.com/9300063/212047815-e1f2a5ed-cf51-4782-b892-66752effe41d.mp4

#### JAXON + SpaceNav in Choreonoid

##### Launch JAXON in Choreonoid

```bash
roscd eus_vive/scripts/jaxon
./start-jaxon_with_rhp3hand-sim.sh
```

```bash
roscd eus_vive/scripts/jaxon
ipython -i jaxon_with_rhp3hand_setup.py
hcf.ast_svc.startAutoBalancer()
hcf.ast_svc.startStabilizer()
hcf.ast_svc.startWholeBodyMasterSlave()
```

##### Launch eus\_vive for SpaceNav

```bash
roslaunch eus_vive jaxon_spacenav_choreonoid.launch
```

### JAXON + Tablis

https://user-images.githubusercontent.com/9300063/212932317-407102d1-093f-4729-b00a-f367bbbcb40d.mp4

#### JAXON + Tablis in Choreonoid

##### Launch Tablis in Choreonoid

```bash
roscd eus_vive/scripts/tablis
./start-tablis-sim.sh
```

```bash
roscd eus_vive/scripts/tablis
ipython -i tablis_setup.py
hcf.servoOn()
hcf.hc_svc.startHapticsController()
```

##### Launch JAXON in Choreonoid

```bash
roscd eus_vive/scripts/jaxon
./start-jaxon_with_rhp3hand-sim.sh
```

```bash
roscd eus_vive/scripts/jaxon
ipython -i jaxon_with_rhp3hand_setup.py
hcf.ast_svc.startAutoBalancer()
hcf.ast_svc.startStabilizer()
hcf.ast_svc.startWholeBodyMasterSlave()
```

##### Launch bridge and wbms core

```bash
roscd eus_vive/scripts/jaxon
./start-jaxon-eus-vive-sim.sh
```

##### Launch eus\_vive for Tablis

```bash
roslaunch eus_vive jaxon_tablis_choreonoid.launch
```

### Demo & Experiments

#### Miraikan Demo 2019/08/23-24

```bash
rossetip
rossetmaster baxter
roslaunch eus_vive baxter_vive_mirror.launch
```

#### Miraikan Demo 2020/09/11-13

##### Robot control PC (Robot side)

```bash
rossetip
rossetmaster baxter
roslaunch eus_vive baxter_remote.launch
```

##### Vive control PC (Pilot side)

```bash
rossetip
rossetmaster baxter
roslaunch eus_vive baxter_vive_remote.launch
```

##### Visualization, display and feedback PC (Pilot side)

```bash
rossetip
rossetmaster baxter
roslaunch eus_vive baxter_display_remote.launch
```

## How to use Vive controller

![Vive controller](https://www.vive.com/media/filer_public/e3/da/e3daf208-4d4e-4adf-b911-22f9458ab883/guid-2d5454b7-1225-449c-b5e5-50a5ea4184d6-web.png)

### PR2

| Button | Usage |
|:-:|:-:|
| 1 / Menu | Control toggle: base/arm (Default: base) |
| 3 / Stream Menu | Steam Menu |
| 8 / Grip | Not used |

#### Arm mode

You can enable arm mode of right and left arm separately.

| Command | Usage |
|:-:|:-:|
| 7 / Trigger | Gripper toggle: open/close in Toggle grasp mode (Default: open) |
| 7 / Trigger | Gripper toggle: open only when trigger is pressed in Hold grasp mode|
| Controller pose | robot end effector's pose |

#### Base mode

Base mode is enabled when both arms are disabled in Arm mode.

| Command | Usage |
|:-:|:-:|
| 2 / Trackpad | Torso control: right: down / left: up |
| 2 / Trackpad + 7 / Trigger (right) | Safe base control: right: x, y / left: w |
| 2 / Trackpad + 7 / Trigger (right + left) | Unsafe base control: right: x, y / left: w |

### Baxter

| Button | Usage |
|:-:|:-:|
| 2 / Trackpad | Control toggle: stop / arm (Default: stop) |
| 3 / Stream Menu | Steam Menu |
| 8 / Grip | Not used |

#### Arm mode

You can enable arm mode of right and left arm separately.

| Command | Usage |
|:-:|:-:|
| 7 / Trigger | Gripper toggle: open/close (Default: open) |
| Controller pose | robot end effector's pose |

## How to use GUI control

### Teleoperation control buttons

| Button | Usage |
|:-:|:-:|
| Reset | Diable and reset robot to initial posture |
| Enable | Enable robot |
| Disable | Disable robot |
| Calibrate larm/rarm | Calibrate with left/right arm |

### Arm & Gripper control buttons

| Button | Usage |
|:-:|:-:|
| Reset larm/rarm | Reset only left/right arm |
| Enable larm/rarm | Enable only left/right arm |
| Disable larm/rarm | Disable only left/right arm |
| Start grasp lgripper/rgripper | Start grasp only left/right gripper |
| Stop grasp lgripper/rgripper | Stop grasp only left/right gripper |
| Toggle grasp mode | Change to Toggle grasp mode |
| Hold grasp mode | Change to Hold grasp mode |

## Tips

### Baxter Network Configuration

Open [Field Service Menu](https://sdk.rethinkrobotics.com/wiki/Field_Service_Menu_(FSM)) and change network configuration

### VPNC Install

```bash
sudo apt install libgcrypt20-dev libgnutls28-dev
git clone https://github.com/streambinder/vpnc.git
cd vpnc
git checkout 1cf24ed6aa4a04b4b01cc9ebfacbad723eed04f5
make
sudo make install
cd ..
git clone git://git.infradead.org/users/dwmw2/vpnc-scripts.git
cd vpnc-scripts
sudo cp vpnc-script /etc/vpnc
```

### VPNC Command for Fortigate

```bash
sudo vpnc --local-port 0 --gateway <gateway> --id ipsecvpn --username <username> --pfs dh5 --dh dh5 --auth-mode psk --no-detach --vendor fortigate --dh dh5
```

## Demo Video

### PR2 Fridge demo

- [Trial 1](https://drive.google.com/open?id=1NsNnplM9bZQi78BY5IJC7CGwROcHd_Nc)
- [Trial 2](https://drive.google.com/open?id=1BSpfWCgKMXQykrTQrrrhYMP5UryFPHr7)
- [Trial 3](https://drive.google.com/open?id=13E_h0JvgZn_RpvnEMDYGR6lNazLvFG0w)

### Baxter APC demo

- [Trial 1](https://drive.google.com/open?id=156y7_M6OwD9z6B4RKwZ0UPLnyHKaNKFm)
- [Trial 2](https://drive.google.com/open?id=1Qb0k0Uzj0pTZhNvM7VIgD6g0RC1ohqg1)
- [Trial 3](https://drive.google.com/open?id=10XZ_5bBKgEk_QqtONCfnXYRLKTE6rpgE)
- [Trial 4](https://drive.google.com/open?id=1IyVME3OggckIfDYCDwIjQmdNneWYdXWd)
- [Trial 5](https://drive.google.com/open?id=1jwq_UdDzgDf-UfBpAS0G0KeykvZ-gLhW)


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package eus_teleop
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.0.10 (2023-04-02)
-------------------
* update readme
* flake8
* disable bash and mypy lint
* flake8
* disable shellcheck
* update readme
* move jaxon tablis scripts in eus_vive
* update tablis launch
* Contributors: Shingo Kitagawa

0.0.9 (2023-01-19)
------------------
* add pr2 and baxter video
* Merge pull request `#47 <https://github.com/knorth55/eus_teleop/issues/47>`_ from knorth55/spacenav-base
* euslint
* updat ereadme
* toggle spacenav base
* fix typo
* add base-cb methods
* euslint
* update controller-button-common-p
* change torso mode
* fix typo
* add controller-button-common-p
* add check torso/base mode functions
* add joy index val
* publish when loop enable
* publish stop base topic
* fix typo
* update jaxon-vr-main.l
* pr2 torso for spacenav
* support spacenav move base
* fix typo in pr2-vr-main.l
* set pr2 spacenav button
* refactor robot-move-base-vr-interface
* refactor main-loop args
* add base button options
* abstract base-cb
* Merge pull request `#46 <https://github.com/knorth55/eus_teleop/issues/46>`_ from knorth55/jaxon
  Add JAXON + SpaceNav, Tablis Demo
* disable md034
* euslint
* add movies
* update readme
* fix typo in readme
* use current-end-coords for spacenav
* refactor jaxon-tablis-interface.l
* fix typo in jaxon-vr-main
* fix typo
* move :reset in :ros-init
* update jaxon tablis params
* add tablis jaxon launch
* add mirror arg in jaxon_spacenav_choreonoid
* change default args in jaxon launch
* load jaxon-tablis-interface in jaxon-vr-main
* add jaxon-tablis-interface
* add feedback pose topic val
* add ros-init
* use prog1 for init and return self in init
* publish leg pose stamp in euslisp
* refactor robot-vr-interface
* add publish-command-pose
* fix typo in robot-vr-interface
* update slot val name
* fix typo
* add always-publish-command-pose
* fix spacenav drifting
* update robot model
* relay leg slave
* refactor robot-vr-interface
* move functions
* update current-end-coords from pose in jaxon
* disable update-current-end-coords in jaxon
* fix topic name
* fix robot-vr-interface visualize
* fix typo
* add advertise
* fix slot search
* support trackpad too
* support other button for base-cb
* check if slot is boundp or not
* fix typo
* rename launches
* add jaxon sample launches
* add jaxon launch files
* add jaxon-vr-main
* publish to command pose topic
* add jaxon custom set reference end coords
* use set-reference-end-coords in robot-vr-interface
* add set-reference-tablis-coords
* refactor parameters
* add arm-cb-solve-ik val
* move update-robot-model to correct position
* add get-target-coords place holder
* add jaxon spacenav interface
* require robot-move-base-spacenav-interface in pr2-spacenav-interface
* fix visualize coords
* add robot-move-base-spacenav-interface.l
* add robot-move-base-oculus-interface.l
* add jaxon-vr-interface
* Merge pull request `#30 <https://github.com/knorth55/eus_teleop/issues/30>`_ from knorth55/dragon-spacenav-demo
* add dragon example in readme
* add dragon.rosinstall
* fix get-target-coords
* add scale
* refactor dragon vr interface
* add dragon vr and gazebo launch
* refactor spacenav euslisp codes
* add dragon-vr-main.l for dragon spacenav demo
* fix typo
* update readme
* update readme
* markdownlint
* update baxter vive sample
* update tablis sample launches
* update readme
* update readme
* Merge pull request `#44 <https://github.com/knorth55/eus_teleop/issues/44>`_ from knorth55/tablis-demo
* fix mirror control
* add mirror arguments
* add tablis.rosinstall
* add tablis bridge follower launch in baxter_vr
* add tablis launches
* add baxter_tablis_remote.launch
* fix tablis get-target-coords
* add scale
* update baxter scale
* use vals for topic names
* call send-super at the end
* refactor parameters
* show rqt gui unless display
* check if device id is set in param
* draw objects in all step
* dont send joy when no diff
* add make-*-irtviewer and camera-model func
* refactor
* euslint
* refactor
* add return t
* fix tablis interface
* comment out pr2 gazebo
* add sound_play
* fix sample
* add move base interface
* refactor launch
* merge pr1012 and pr1040 launch
* use require
* change org method names
* overwrite start-grasp and stop-grasp
* add tablis in main
* add baxter and pr2 launches
* add baxter tablis interface
* add pr2 tablis interface
* add robot-tablis-interface
* add update-robot-model method
* update baxter_tabletop_object_detector.launch
* fix baxter_logging.launch
* fix baxter.launch
* add allow-other-keys t for baxter-init
* fix baxter-init &rest args
* Contributors: Shingo Kitagawa

0.0.8 (2022-08-15)
------------------
* Merge pull request `#43 <https://github.com/knorth55/eus_teleop/issues/43>`_ from Kanazawanaoaki/add-moveit-to-depend
* Merge pull request `#42 <https://github.com/knorth55/eus_teleop/issues/42>`_ from Kanazawanaoaki/add-eus_teleop-to-rosinstall
* Add pr2_moveit_config to depend in package.xml
* Add eus_teleop to rosinstall
* Merge pull request `#41 <https://github.com/knorth55/eus_teleop/issues/41>`_ from knorth55/baxter-moveit-args
  add gripper args in baxter_moveit.launch
* add gripper args in baxter_moveit.launch
* Merge pull request `#40 <https://github.com/knorth55/eus_teleop/issues/40>`_ from knorth55/use-baxter-softhand-interface
  use baxter-softhand-interface in jsk_robot
* use baxter-softhand-interface in jsk_robot
* Merge pull request `#37 <https://github.com/knorth55/eus_teleop/issues/37>`_ from YUKINA-3252/global-time
* Make global_time_enabled true
* Update README.md
* update rviz config
* update moveit rviz
* update rviz
* update rviz config
* add allow-other-keys
* update *baxter* with potentio vector
* add finger1,2,3 rotate method in baxter-interface
* Merge pull request `#36 <https://github.com/knorth55/eus_teleop/issues/36>`_ from softyanija/update-rotate-angle
* add rotate-angle in baxter-interface.l
* update baxter.rosinstall

scripts/tablis/tablis_setup.py
#!/usr/bin/env python

import sys
import time

from hrpsys import rtm
rtm.nshost = "localhost"
rtm.nsport = "15006"
rtm.initCORBA()

from hrpsys.OpenHRP import *  # NOQA
import OpenHRP  # NOQA

from haptics_controller import *  # NOQA
from haptics_controller.SimpleHapticsControllerService_idl import *  # NOQA


def findComp(name):
    timeout_count = 0
    comp = None
    while timeout_count < 10:
        comp = rtm.findRTC(name)
        if comp is not None and comp.isActive():
            break
        print("find Comp wait for " + name)
        time.sleep(1)
        timeout_count += 1
    if comp is None:
        print("Cannot find component: %s" % name)
    return comp


class TABLIS_Configurator:
    Groups = {
        'rleg': [
            'RLEG_JOINT0',
            'RLEG_JOINT1',
            'RLEG_JOINT2',
            'RLEG_JOINT3',
            'RLEG_JOINT4',
            'RLEG_JOINT5'
        ],
        'lleg': [
            'LLEG_JOINT0',
            'LLEG_JOINT1',
            'LLEG_JOINT2',
            'LLEG_JOINT3',
            'LLEG_JOINT4',
            'LLEG_JOINT5'
        ],
        'torso': [
            'CHEST_JOINT0'
        ],
        'rarm': [
            'RARM_JOINT0',
            'RARM_JOINT1',
            'RARM_JOINT2',
            'RARM_JOINT3',
            'RARM_JOINT4',
            'RARM_JOINT5',
            'RARM_JOINT6',
            'RARM_JOINT7'
        ],
        'larm': [
            'LARM_JOINT0',
            'LARM_JOINT1',
            'LARM_JOINT2',
            'LARM_JOINT3',
            'LARM_JOINT4',
            'LARM_JOINT5',
            'LARM_JOINT6',
            'LARM_JOINT7'
        ]}

    rh_svc = rtm.findService(
        findComp("RobotHardware0"),
        "RobotHardwareService",
        "RobotHardwareService",
        "service0"
    )._narrow(OpenHRP.RobotHardwareService)
    seq_svc = rtm.findService(
        findComp("seq"),
        "SequencePlayerService",
        "SequencePlayerService",
        "service0"
    )._narrow(OpenHRP.SequencePlayerService)
    sh_svc = rtm.findService(
        findComp("sh"),
        "StateHolderService",
        "StateHolderService",
        "service0"
    )._narrow(OpenHRP.StateHolderService)
    hc_svc = rtm.findService(
        findComp("hc"),
        "SimpleHapticsControllerService",
        "SimpleHapticsControllerService",
        "service0"
    )._narrow(OpenHRP.SimpleHapticsControllerService)
    log_svc = findComp("log").service("service0")._narrow(
        OpenHRP.DataLoggerService)

    def servoOn(self, jname='all', tm=3):
        self.servoOff()

        c = False
        while (c != 'Y' and c != 'y'):
            c = raw_input("press 'Y' for servo ON and power ON. >> ")

        # urata system flag resetting for jaxon
        self.rh_svc.power(jname, OpenHRP.RobotHardwareService.SWITCH_ON)
        time.sleep(0.01)
        self.rh_svc.power(jname, OpenHRP.RobotHardwareService.SWITCH_OFF)
        time.sleep(0.02)

        self.rh_svc.setServoGainPercentage(jname, 0.0)
        self.rh_svc.setServoTorqueGainPercentage(jname, 100)

        # reset JointGroups
        for k, v in self.Groups.items():
            self.seq_svc.removeJointGroup(k)
        for k, v in self.Groups.items():
            self.seq_svc.waitInterpolationOfGroup(k)
        for k, v in self.Groups.items():
            self.seq_svc.addJointGroup(k, v)

        # move to idle mode for filter type RTCs
        for rtc_name in ["hc"]:
            rtc = rtm.findRTC(rtc_name)
            if rtc:
                rtc.stop()
                rtc.start()

        time.sleep(2)
        self.sh_svc.goActual()
        self.seq_svc.setBasePos([0, 0, 1.44], 0.01)
        time.sleep(0.1)
        self.rh_svc.power(jname, OpenHRP.RobotHardwareService.SWITCH_ON)
        time.sleep(0.1)
        self.rh_svc.servo(jname, OpenHRP.RobotHardwareService.SWITCH_ON)
        return True

    def servoOff(self, jname='all'):
        c = False
        while (c != 'Y' and c != 'y'):
            c = raw_input("press 'Y' for servo OFF and power OFF. >> ")

        self.rh_svc.servo('all', OpenHRP.RobotHardwareService.SWITCH_OFF)
        time.sleep(0.2)
        if jname == 'all':
            self.rh_svc.power('all', OpenHRP.RobotHardwareService.SWITCH_OFF)
        return True

    def setHcParametersJAXON(self):
        # ast setting
        hcp = self.hc_svc.getHapticsControllerParam()[1]
        hcp.max_torque = [
            20, 80, 80, 80, 20, 20,
            20, 80, 80, 80, 20, 20,
            40,
            30, 30, 20, 20, 12, 8, 8, 6,
            30, 30, 20, 20, 12, 8, 8, 6
        ]
        hcp.qref_dgain = 10.0  # simulator only
        self.hc_svc.setHapticsControllerParam(hcp)

    def setupLogger(self):
        self.log_svc.add("TimedDoubleSeq", "RobotHardware0_q")
        rtm.connectPorts(
            rtm.findRTC("RobotHardware0").port("q"),
            rtm.findRTC("log").port("RobotHardware0_q")
        )
        self.log_svc.add("TimedDoubleSeq", "RobotHardware0_dq")
        rtm.connectPorts(
            rtm.findRTC("RobotHardware0").port("dq"),
            rtm.findRTC("log").port("RobotHardware0_dq")
        )
        self.log_svc.add("TimedDoubleSeq", "RobotHardware0_tau")
        rtm.connectPorts(
            rtm.findRTC("RobotHardware0").port("tau"),
            rtm.findRTC("log").port("RobotHardware0_tau")
        )
        self.log_svc.add("TimedLongSeqSeq", "RobotHardware0_servoState")
        rtm.connectPorts(
            rtm.findRTC("RobotHardware0").port("servoState"),
            rtm.findRTC("log").port("RobotHardware0_servoState")
        )
        self.log_svc.add("TimedDoubleSeq", "hc_genTauOut")
        rtm.connectPorts(
            rtm.findRTC("hc").port("genTauOut"),
            rtm.findRTC("log").port("hc_genTauOut")
        )
        for ee in ["rleg", "lleg", "rarm", "larm", "relbow", "lelbow"]:
            self.log_svc.add("TimedPose3D", "hc_act" + ee + "PoseOut")
            rtm.connectPorts(
                rtm.findRTC("hc").port("act" + ee + "PoseOut"),
                rtm.findRTC("log").port("hc_act" + ee + "PoseOut")
            )
        self.log_svc.maxLength(1000 * 60)
        self.log_svc.clear()

    def init(self):
        for j in self.Groups["rleg"] + self.Groups["lleg"] \
                + self.Groups["torso"] + self.Groups["rarm"] \
                + self.Groups["larm"]:
            self.rh_svc.setJointControlMode(
                j, OpenHRP.RobotHardwareService.TORQUE)
            self.rh_svc.setServoErrorLimit(j, 0.0)
            self.rh_svc.setServoTorqueGainPercentage(j, 100)
            self.rh_svc.setServoGainPercentage(j, 0.0)

        self.setupLogger()
        self.seq_svc.setBasePos([0, 0, 1.44], 2.0)
        self.setHcParametersJAXON()


if __name__ == '__main__':
    hcf = TABLIS_Configurator()

    if len(sys.argv) > 1 and sys.argv[1] == "init":
        hcf.init()
